worker/src/direct_invocation/ws_handler.rs

```diff
@@ -0,0 +1,157 @@
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+struct DirectWsServerRequest {
+    pub client: Sender,
+    pub request: String,
+}
+
+impl DirectWsServerRequest {
+    pub fn new(client: Sender, request: String) -> Self {
+        Self { client, request }
+    }
+}
+
+/// WebSocket Handler implementation
+pub struct WsHandler<E, W> {
+    client: Sender,
+    enclave_api: E,
+    watch_list: Arc<W>,
+}
+
+impl<E, W> Handler for WsHandler<E, W>
+where
+    E: EnclaveApi,
+    W: WatchListFacade,
+{
+    fn on_message(&mut self, msg: Message) -> Result<()> {
+        let request = DirectWsServerRequest::new(self.client.clone(), msg.to_string());
+        if self.handle_direct_invocation_request(request).is_err() {
+            error!("direct invocation call was not successful");
+        }
+        Ok(())
+    }
+
+    fn on_close(&mut self, code: CloseCode, reason: &str) {
+        debug!(
+            "Direct invocation WebSocket closing for ({:?}) {}",
+            code, reason
+        );
+    }
+}
+
+impl<E, W> WsHandler<E, W>
+where
+    E: EnclaveApi,
+    W: WatchListFacade,
+{
+    fn handle_direct_invocation_request(&self, req: DirectWsServerRequest) -> Result<()> {
+        info!("Got message '{:?}'. ", req.request);
+
+        let msg: Vec<u8> = req.request.as_bytes().to_vec();
+
+        let response = self
+            .enclave_api
+            .rpc(msg)
+            .map_err(|e| ws::Error::new(ws::ErrorKind::Internal, format!("{:?}", e)))?;
+
+        let decoded_response = String::from_utf8_lossy(&response).to_string();
+        if let Ok(full_rpc_response) =
+            serde_json::from_str(&decoded_response) as serde_json::Result<RpcResponse>
+        {
+            if let Ok(result_of_rpc_response) =
+                RpcReturnValue::decode(&mut full_rpc_response.result.as_slice())
+            {
+                if let DirectRequestStatus::TrustedOperationStatus(_) =
+                    result_of_rpc_response.status
+                {
+                    if result_of_rpc_response.do_watch {
+                        // start watching the call with the specific hash
+                        if let Ok(hash) = Hash::decode(&mut result_of_rpc_response.value.as_slice())
+                        {
+                            // create new key and value entries to store
+                            let new_client = WatchingClient {
+                                client: req.client.clone(),
+                                response: RpcResponse {
+                                    result: result_of_rpc_response.encode(),
+                                    jsonrpc: full_rpc_response.jsonrpc.clone(),
+                                    id: full_rpc_response.id,
+                                },
+                            };
+
+                            self.watch_list.add_watching_client(hash, new_client);
+                        }
+                    }
+                }
+            }
```

Comment:
I think this code is too nested for something that is actually a simple linear flow. We can replace the nested stuff with some simple `?`, when we introduce the appropriate Error(-Conversions).

Reply:
Agreed, I will try to refactor this section and make it more readable. The outermost scope can be replaced by `?`, but the inner ones do not simply return when an error occurs, rather it only executes that action when all the conditions are met. Regardless of that outcome however, it will always use `client.send()` - so we cannot use `?`, because we must not return.

Reply:
Ok, agreed we can't get rid of all the nesting, but it is not only the outermost scope, there are some `decode` functions that can also be replaced: L122, L125, L133